codegen/frameCodegen.ts

```diff
@@ -0,0 +1,199 @@
+    }
+
+    for (const constant of Object.values(pallet.constants)) {
+      palletDeclarationStatements.push(`${constant.name}: ${typeCodegen.native(constant.codec)}`)
+      palletStatements.push(
+        `${constant.name} = ${codecCodegen.print(constant.codec)}.decode(C.hex.decode("${
+          hex.encode(constant.value)
+        }"))`,
+      )
+    }
+
+    const { call } = pallet.types
+    if (call) {
+      const factory = typeCodegen.typeNames.get(call)!
+      new $.CodecVisitor<void>()
+        .add(
+          $.taggedUnion,
+          (_codec, _tagKey: string, variantsRaw) =>
+            (Array.isArray(variantsRaw) ? variantsRaw : Object.values(variantsRaw)).forEach(
+              (variant) => {
+                palletDeclarationStatements.push(
+                  `
+                    ${variant.tag}: <X>(
+                      props: C.RunicArgs<X, Omit<${factory}.${variant.tag}, "type">>
+                    ) => C.ExtrinsicRune<${chainIdent}, C.RunicArgs.U<X>>
+                  `,
+                )
+                palletStatements.push(extrinsicFactory(variant.tag))
+              },
+            ),
+        )
+        .add($.literalUnion<unknown>, (_, values) =>
+          Object.values(values).forEach((value) => {
+            if (typeof value !== "string") {
+              throw new Error("pallet call non-string literalUnion is unsupported")
+            }
+            palletDeclarationStatements.push(
+              `${value}: () => C.ExtrinsicRune<${chainIdent}, never>`,
+            )
+            palletStatements.push(extrinsicFactory(value))
+          }))
+        .add($.never, () => {})
+        .visit(call)
+
+      // deno-lint-ignore no-inner-declarations
+      function extrinsicFactory(methodIdent: string) {
+        return `
+          ${methodIdent} = (...args) => this.chain.extrinsic(C.Rune.object({
+            type: "${pallet.name}",
+            value: ${factory}.${methodIdent}(...args),
+          }))
+        `
+      }
+    }
+
+    palletDeclarations.push(`
+      export class ${pallet.name}PalletRune<out U> extends C.PalletRune<${chainIdent}, "${pallet.name}", U> {
+        ${palletDeclarationStatements.join("\n")}
+      }
+    `)
+    palletDefinitions.push(`
+      export class ${pallet.name}PalletRune extends C.PalletRune {
+        ${palletStatements.join("\n")}
+      }
+    `)
+  }
+
+  files.set(
+    "pallets.d.ts",
+    `
+      ${IMPORTS_COMMON}
+      ${palletDeclarations.join("\n")}
+    `,
+  )
+  files.set(
+    "pallets.js",
+    `
+      ${IMPORTS_COMMON}
+      ${palletDefinitions.join("\n")}
+    `,
+  )
+
+  files.set(
+    "chain.d.ts",
+    `
+      ${IMPORTS_COMMON}
+      import { metadata } from "./metadata.js"
+
+      export interface ${chainIdent} extends C.Chain<metadata> {}
+
+      export class ${chainIdent}ChainRune<U> extends C.ChainRune<${chainIdent}, U> {
+        static override from(connect: (signal: AbortSignal) => C.Connection): ${chainIdent}ChainRune<U>
```

Comment:
`U` isn't defined at the static level
```suggestion
        static override from(connect: (signal: AbortSignal) => C.Connection): ${chainIdent}ChainRune<never>
```